File: dotnet/PolymarketBot/Services/Estimator.cs
using System.Text;
using System.Text.Json;
using Microsoft.Extensions.Logging;
using PolymarketBot.Models;

namespace PolymarketBot.Services;

public sealed class Estimator
{
    private const string SystemPrompt =
        "You are a calibrated probability estimator for prediction markets.\n" +
        "Given a market question, estimate the TRUE probability that the outcome resolves YES.\n" +
        "\n" +
        "Rules:\n" +
        "- Output ONLY valid JSON: {\"probability\": 0.XX, \"reasoning\": \"one sentence\"}\n" +
        "- probability must be between 0.02 and 0.98\n" +
        "- Be well-calibrated: events you rate at 70% should happen ~70% of the time\n" +
        "- Use base rates, current knowledge, and logical reasoning\n" +
        "- The current market price reflects real-money consensus from many informed traders — treat it as a Bayesian prior. Only deviate significantly if you have strong specific reasoning.\n" +
        "- If deeply uncertain, stay close to the market price\n" +
        "- Keep reasoning under 50 words";

    private readonly BotConfig _config;
    private readonly HttpClient _http;
    private readonly ILogger<Estimator> _log;

    public Estimator(BotConfig config, HttpClient http, ILogger<Estimator> log)
    {
        _config = config;
        _http = http;
        _log = log;
    }

    public async Task<Estimate?> EstimateAsync(MarketInfo market, CancellationToken ct = default)
    {
        var rawEstimates = new List<double>();
        var totalInput = 0;
        var totalOutput = 0;
        var firstReasoning = "";

        for (var i = 0; i < _config.EnsembleSize; i++)
        {
            var result = await SingleCallAsync(market, ct);
            if (result is null) continue;

            rawEstimates.Add(result.Value.Probability);
            if (string.IsNullOrEmpty(firstReasoning))
                firstReasoning = result.Value.Reasoning;
            totalInput += result.Value.InputTokens;
            totalOutput += result.Value.OutputTokens;
        }

        if (rawEstimates.Count < 2)
        {
            _log.LogWarning("Only {Count} valid estimates for: {Question}",
                rawEstimates.Count, Truncate(market.Question, 60));
            if (rawEstimates.Count == 0) return null;
        }

        // Trimmed mean: drop highest and lowest if enough samples
        List<double> trimmed;
        if (rawEstimates.Count >= 4)
        {
            var sorted = rawEstimates.OrderBy(x => x).ToList();
            trimmed = sorted.Skip(1).Take(sorted.Count - 2).ToList();
        }
        else
        {
            trimmed = rawEstimates;
        }

        var fairProb = trimmed.Average();
        var confidence = rawEstimates.Count > 1 ? StdDev(rawEstimates) : 1.0;

        // Confidence filter: skip if ensemble disagreement is too high
        if (rawEstimates.Count >= 2 && confidence > _config.MaxEstimateStd)
        {
            _log.LogInformation("SKIP (low confidence): {Question} std={Std:F3} > max={Max:F3}",
                Truncate(market.Question, 50), confidence, _config.MaxEstimateStd);
            return null;
        }

        _log.LogInformation("Estimate: {Question} -> {Prob:P2} (n={Count}, std={Std:F3})",
            Truncate(market.Question, 50), fairProb, rawEstimates.Count, confidence);

        return new Estimate
        {
            MarketConditionId = market.ConditionId,
            Question = market.Question,
            FairProbability = fairProb,
            RawEstimates = rawEstimates,
            Confidence = confidence,
            ReasoningSummary = firstReasoning,
            InputTokensUsed = totalInput,
            OutputTokensUsed = totalOutput,
        };
    }

    private async Task<CallResult?> SingleCallAsync(MarketInfo market, CancellationToken ct)
    {
        // Retry delays for transient overload (429 / 529): 10s, 20s, 40s
        int[] backoffMs = { 10_000, 20_000, 40_000 };

        for (var attempt = 0; attempt <= backoffMs.Length; attempt++)
        {
            try
            {
                var (status, body) = await MakeProviderRequestAsync(market, ct);

                if (status == 429 || status == 529)
                {
                    if (attempt < backoffMs.Length)
                    {
                        var delay = backoffMs[attempt];
                        _log.LogWarning("{Provider} {Status} (attempt {A}/{Max}) — retrying in {Sec}s",
                            _config.AiProvider, status, attempt + 1, backoffMs.Length, delay / 1000);
                        await Task.Delay(delay, ct);
                        continue;
                    }
                    _log.LogError("{Provider} {Status}: giving up after {Max} retries for {Question}",
                        _config.AiProvider, status, backoffMs.Length, Truncate(market.Question, 40));
                    return null;
                }

                if (status < 200 || status >= 300)
                {
                    _log.LogError("{Provider} HTTP {Status}: {Body}",
                        _config.AiProvider, status, body[..Math.Min(body.Length, 200)]);
                    return null;
                }

                return ParseProviderResponse(body);
            }
            catch (JsonException ex)
            {
                _log.LogDebug("Failed to parse estimate response: {Error}", ex.Message);
                return null;
            }
            catch (Exception ex) when (ex is not OperationCanceledException)
            {
                _log.LogDebug("{Provider} call failed: {Error}", _config.AiProvider, ex.Message);
                return null;
            }
        }

        return null;
    }

    private string ActiveModel =>
        !string.IsNullOrEmpty(_config.AiModel) ? _config.AiModel : _config.ClaudeModel;

    private async Task<(int status, string body)> MakeProviderRequestAsync(MarketInfo market, CancellationToken ct)
    {
        var userPrompt = BuildUserPrompt(market);
        return _config.AiProvider.ToLowerInvariant() switch
        {
            "gemini"       => await MakeGeminiRequestAsync(userPrompt, ct),
            "openai"       => await MakeOpenAiCompatRequestAsync("openai", userPrompt, ct),
            "openrouter"   => await MakeOpenAiCompatRequestAsync("openrouter", userPrompt, ct),
            "azure_openai" => await MakeOpenAiCompatRequestAsync("azure_openai", userPrompt, ct),
            _              => await MakeAnthropicRequestAsync(userPrompt, ct),  // default: anthropic
        };
    }

    // ── Anthropic ──────────────────────────────────────────────────────────

    private async Task<(int, string)> MakeAnthropicRequestAsync(string userPrompt, CancellationToken ct)
    {
        var body = JsonSerializer.Serialize(new
        {
            model = ActiveModel,
            max_tokens = _config.MaxEstimateTokens,
            temperature = _config.EnsembleTemperature,
            system = SystemPrompt,
            messages = new[] { new { role = "user", content = userPrompt } }
        });
        var host = string.IsNullOrEmpty(_config.AnthropicApiHost) ? "https://api.anthropic.com" : _config.AnthropicApiHost;
        var req = new HttpRequestMessage(HttpMethod.Post, $"{host}/v1/messages")
        {
            Content = new StringContent(body, Encoding.UTF8, "application/json")
        };
        req.Headers.Add("x-api-key", _config.AnthropicApiKey);
        req.Headers.Add("anthropic-version", "2023-06-01");
        var resp = await _http.SendAsync(req, ct);
        return ((int)resp.StatusCode, await resp.Content.ReadAsStringAsync(ct));
    }

    private static CallResult? ParseAnthropicResponse(string body)
    {
        var doc = JsonDocument.Parse(body);
        var text = doc.RootElement.GetProperty("content")[0].GetProperty("text").GetString()?.Trim() ?? "";
        var usage = doc.RootElement.GetProperty("usage");
        var inputTokens = usage.GetProperty("input_tokens").GetInt32();
        var outputTokens = usage.GetProperty("output_tokens").GetInt32();
        var (prob, reasoning) = ParseProbabilityJson(text);
        return prob < 0 ? null : new CallResult(prob, reasoning, inputTokens, outputTokens);
    }

    // ── OpenAI-compatible (OpenAI, OpenRouter, Azure OpenAI) ──────────────

    private async Task<(int, string)> MakeOpenAiCompatRequestAsync(string provider, string userPrompt, CancellationToken ct)
    {
        string url;
        var model = ActiveModel;
        var req = new HttpRequestMessage(HttpMethod.Post, "");

        if (provider == "openai")
        {
            var host = string.IsNullOrEmpty(_config.OpenAiApiHost) ? "https://api.openai.com" : _config.OpenAiApiHost.TrimEnd('/');
            url = $"{host}/v1/chat/completions";
            req.Headers.Add("Authorization", $"Bearer {_config.OpenAiApiKey}");
        }
        else if (provider == "openrouter")
        {
            url = "https://openrouter.ai/api/v1/chat/completions";
            req.Headers.Add("Authorization", $"Bearer {_config.OpenRouterApiKey}");
        }
        else  // azure_openai
        {
            var endpoint = _config.AzureOpenAiEndpoint.TrimEnd('/');
            var deployment = _config.AzureOpenAiDeployment;
            var version = string.IsNullOrEmpty(_config.AzureOpenAiApiVersion) ? "2024-02-01" : _config.AzureOpenAiApiVersion;
            url = $"{endpoint}/openai/deployments/{deployment}/chat/completions?api-version={version}";
            req.Headers.Add("api-key", _config.AzureOpenAiApiKey);
            model = deployment;  // Azure uses deployment name
        }

        var body = JsonSerializer.Serialize(new
        {
            model,
            messages = new[]
            {
                new { role = "system", content = SystemPrompt },
                new { role = "user",   content = userPrompt }
            },
            temperature = _config.EnsembleTemperature,
            max_tokens = _config.MaxEstimateTokens,
        });
        req.RequestUri = new Uri(url);
        req.Content = new StringContent(body, Encoding.UTF8, "application/json");
        var resp = await _http.SendAsync(req, ct);
        return ((int)resp.StatusCode, await resp.Content.ReadAsStringAsync(ct));
    }

    private static CallResult? ParseOpenAiCompatResponse(string body)
    {
        var doc = JsonDocument.Parse(body);
        var text = doc.RootElement
            .GetProperty("choices")[0]
            .GetProperty("message")
            .GetProperty("content")
            .GetString()?.Trim() ?? "";
        var usage = doc.RootElement.TryGetProperty("usage", out var u) ? u : (JsonElement?)null;
        var inputTokens  = usage?.TryGetProperty("prompt_tokens",     out var pt) == true ? pt.GetInt32() : 0;
        var outputTokens = usage?.TryGetProperty("completion_tokens", out var ct2) == true ? ct2.GetInt32() : 0;
        var (prob, reasoning) = ParseProbabilityJson(text);
        return prob < 0 ? null : new CallResult(prob, reasoning, inputTokens, outputTokens);
    }

    // ── Google Gemini ──────────────────────────────────────────────────────

    private async Task<(int, string)> MakeGeminiRequestAsync(string userPrompt, CancellationToken ct)
    {
        var model = ActiveModel;
        var url = $"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={_config.GeminiApiKey}";
        var body = JsonSerializer.Serialize(new
        {
            systemInstruction = new { parts = new[] { new { text = SystemPrompt } } },
            contents = new[] { new { role = "user", parts = new[] { new { text = userPrompt } } } },
            generationConfig = new
            {
                temperature = _config.EnsembleTemperature,
                maxOutputTokens = _config.MaxEstimateTokens,
            }
        });
        var req = new HttpRequestMessage(HttpMethod.Post, url)
        {
            Content = new StringContent(body, Encoding.UTF8, "application/json")
        };
        var resp = await _http.SendAsync(req, ct);
        return ((int)resp.StatusCode, await resp.Content.ReadAsStringAsync(ct));
    }

    private static CallResult? ParseGeminiResponse(string body)
    {
        var doc = JsonDocument.Parse(body);
        var text = doc.RootElement
            .GetProperty("candidates")[0]
            .GetProperty("content")
            .GetProperty("parts")[0]
            .GetProperty("text")
            .GetString()?.Trim() ?? "";
        var meta = doc.RootElement.TryGetProperty("usageMetadata", out var m) ? m : (JsonElement?)null;
        var inputTokens  = meta?.TryGetProperty("promptTokenCount",     out var pt) == true ? pt.GetInt32() : 0;
        var outputTokens = meta?.TryGetProperty("candidatesTokenCount", out var ct2) == true ? ct2.GetInt32() : 0;
        var (prob, reasoning) = ParseProbabilityJson(text);
        return prob < 0 ? null : new CallResult(prob, reasoning, inputTokens, outputTokens);
    }

    // ── Response parsing (shared) ──────────────────────────────────────────

    private CallResult? ParseProviderResponse(string body)
    {
        return _config.AiProvider.ToLowerInvariant() switch
        {
            "gemini"       => ParseGeminiResponse(body),
            "openai"       => ParseOpenAiCompatResponse(body),
            "openrouter"   => ParseOpenAiCompatResponse(body),
            "azure_openai" => ParseOpenAiCompatResponse(body),
            _              => ParseAnthropicResponse(body),
        };
    }

    /// <summary>
    /// Parse {"probability": 0.XX, "reasoning": "..."} from model response text.
    /// Returns (prob, reasoning) where prob = -1 signals parse failure.
    /// </summary>
    private static (double prob, string reasoning) ParseProbabilityJson(string text)
    {
        try
        {
            if (text.StartsWith("```"))
            {
                var lines = text.Split('\n').Where(l => !l.TrimStart().StartsWith("```")).ToArray();
                text = string.Join('\n', lines);
            }
            var doc = JsonDocument.Parse(text);
            var prob = doc.RootElement.GetProperty("probability").GetDouble();
            var reasoning = doc.RootElement.TryGetProperty("reasoning", out var r) ? r.GetString() ?? "" : "";
            return (Math.Clamp(prob, 0.02, 0.98), reasoning);
        }
        catch
        {
            return (-1, "");
        }
    }

    private static string BuildUserPrompt(MarketInfo market)
    {
        var desc = market.Description.Length > 500
            ? market.Description[..500]
            : market.Description;
        if (string.IsNullOrEmpty(desc)) desc = "N/A";

        return $"Market: {market.Question}\n" +
            $"Event: {market.EventTitle}\n" +
            $"Description: {desc}\n" +
            $"Category: {market.Category}\n" +
            $"Resolution date: {(string.IsNullOrEmpty(market.EndDate) ? "Unknown" : market.EndDate)}\n" +
            $"Current market price: YES at {market.OutcomeYesPrice:P0} / NO at {market.OutcomeNoPrice:P0}\n" +
            "\n" +
            "Estimate the probability this resolves YES. Output JSON only.";
    }

    private static double StdDev(List<double> values)
    {
        var mean = values.Average();
        var sumSq = values.Sum(v => (v - mean) * (v - mean));
        return Math.Sqrt(sumSq / (values.Count - 1));
    }

    /// <summary>
    /// Makes a minimal test call to validate the configured provider's API key.
    /// Returns false on auth failure (HTTP 401/403). Network/rate-limit errors return true (don't block startup).
    /// </summary>
    public async Task<bool> ValidateApiKeyAsync(CancellationToken ct = default)
    {
        try
        {
            var provider = _config.AiProvider.ToLowerInvariant();
            HttpRequestMessage req;

            if (provider == "gemini")
            {
                var model = ActiveModel;
                var url = $"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={_config.GeminiApiKey}";
                var body = JsonSerializer.Serialize(new
                {
                    contents = new[] { new { parts = new[] { new { text = "hi" } } } },
                    generationConfig = new { maxOutputTokens = 1 }
                });
                req = new HttpRequestMessage(HttpMethod.Post, url)
                {
                    Content = new StringContent(body, Encoding.UTF8, "application/json")
                };
            }
            else if (provider is "openai" or "openrouter" or "azure_openai")
            {
                var (url, authHeader, authValue) = provider switch
                {
                    "openai" => ($"{(_config.OpenAiApiHost.TrimEnd('/'))}/v1/chat/completions", "Authorization", $"Bearer {_config.OpenAiApiKey}"),
                    "openrouter" => ("https://openrouter.ai/api/v1/chat/completions", "Authorization", $"Bearer {_config.OpenRouterApiKey}"),
                    _ => ($"{_config.AzureOpenAiEndpoint.TrimEnd('/')}/openai/deployments/{_config.AzureOpenAiDeployment}/chat/completions?api-version={_config.AzureOpenAiApiVersion}", "api-key", _config.AzureOpenAiApiKey),
                };
                var body = JsonSerializer.Serialize(new
                {
                    model = provider == "azure_openai" ? _config.AzureOpenAiDeployment : ActiveModel,
                    messages = new[] { new { role = "user", content = "hi" } },
                    max_tokens = 1
                });
                req = new HttpRequestMessage(HttpMethod.Post, url)
                {
                    Content = new StringContent(body, Encoding.UTF8, "application/json")
                };
                req.Headers.Add(authHeader, authValue);
            }
            else  // anthropic
            {
                var host = string.IsNullOrEmpty(_config.AnthropicApiHost) ? "https://api.anthropic.com" : _config.AnthropicApiHost;
                var body = JsonSerializer.Serialize(new
                {
                    model = ActiveModel,
                    max_tokens = 1,
                    messages = new[] { new { role = "user", content = "hi" } }
                });
                req = new HttpRequestMessage(HttpMethod.Post, $"{host}/v1/messages")
                {
                    Content = new StringContent(body, Encoding.UTF8, "application/json")
                };
                req.Headers.Add("x-api-key", _config.AnthropicApiKey);
                req.Headers.Add("anthropic-version", "2023-06-01");
            }

            var resp = await _http.SendAsync(req, ct);
            var status = (int)resp.StatusCode;

            if (status == 401 || status == 403)
            {
                var respBody = await resp.Content.ReadAsStringAsync(ct);
                _log.LogError("{Provider} API key invalid (HTTP {Status}): {Body}",
                    _config.AiProvider, status, respBody[..Math.Min(respBody.Length, 200)]);
                return false;
            }
            // Gemini returns 400 for invalid keys — check response body
            if (provider == "gemini" && status == 400)
            {
                var respBody = await resp.Content.ReadAsStringAsync(ct);
                if (respBody.Contains("API key", StringComparison.OrdinalIgnoreCase))
                {
                    _log.LogError("Gemini API key invalid: {Body}", respBody[..Math.Min(respBody.Length, 200)]);
                    return false;
                }
            }
            return true;
        }
        catch (Exception ex) when (ex is not OperationCanceledException)
        {
            _log.LogWarning("API key validation network error: {Error} — continuing", ex.Message);
            return true;
        }
    }

    private static string Truncate(string s, int maxLen)
        => s.Length <= maxLen ? s : s[..maxLen] + "...";

    private readonly record struct CallResult(double Probability, string Reasoning, int InputTokens, int OutputTokens);
}

File: python/estimator.py
"""AI ensemble probability estimation for prediction markets.

Supported providers: anthropic, openai, openrouter, azure_openai, gemini
"""

import json
import logging
import statistics
import time
from typing import Optional

import requests

from config import BotConfig
from models import MarketInfo, Estimate

log = logging.getLogger("bot.estimator")

SYSTEM_PROMPT = """You are a calibrated probability estimator for prediction markets.
Given a market question, estimate the TRUE probability that the outcome resolves YES.

Rules:
- Output ONLY valid JSON: {"probability": 0.XX, "reasoning": "one sentence"}
- probability must be between 0.02 and 0.98
- Be well-calibrated: events you rate at 70% should happen ~70% of the time
- Use base rates, current knowledge, and logical reasoning
- The current market price reflects real-money consensus from many informed traders — treat it as a Bayesian prior. Only deviate significantly if you have strong specific reasoning.
- If deeply uncertain, stay close to the market price
- Keep reasoning under 50 words"""


def _build_user_prompt(market: MarketInfo) -> str:
    desc = market.description[:500] if market.description else "N/A"
    return (
        f"Market: {market.question}\n"
        f"Event: {market.event_title}\n"
        f"Description: {desc}\n"
        f"Category: {market.category}\n"
        f"Resolution date: {market.end_date or 'Unknown'}\n"
        f"Current market price: YES at {market.outcome_yes_price:.0%} / NO at {market.outcome_no_price:.0%}\n\n"
        f"Estimate the probability this resolves YES. Output JSON only."
    )


class Estimator:
    def __init__(self, config: BotConfig):
        self.config = config
        self._provider = config.ai_provider.lower()
        self._model = config.ai_model or config.claude_model  # backward compat

        if self._provider == "anthropic":
            import anthropic as _anthropic_sdk
            self._anthropic_sdk = _anthropic_sdk
            kwargs: dict = {"api_key": config.anthropic_api_key}
            if config.anthropic_api_host:
                kwargs["base_url"] = config.anthropic_api_host
            self._anthropic_client = _anthropic_sdk.Anthropic(**kwargs)
        else:
            self._anthropic_sdk = None
            self._anthropic_client = None

    def estimate(self, market: MarketInfo) -> Optional[Estimate]:
        """Run ensemble estimation: N independent AI calls, trimmed mean."""
        raw_estimates: list[float] = []
        total_input = 0
        total_output = 0
        first_reasoning = ""

        for _ in range(self.config.ensemble_size):
            result = self._single_call(market)
            if result is None:
                continue
            prob, reasoning, in_tok, out_tok = result
            raw_estimates.append(prob)
            if not first_reasoning:
                first_reasoning = reasoning
            total_input += in_tok
            total_output += out_tok

        if len(raw_estimates) < 2:
            log.warning(f"Only {len(raw_estimates)} valid estimates for: {market.question[:60]}")
            if not raw_estimates:
                return None

        if len(raw_estimates) >= 4:
            sorted_est = sorted(raw_estimates)
            trimmed = sorted_est[1:-1]
        else:
            trimmed = raw_estimates

        fair_prob = statistics.mean(trimmed)
        confidence = statistics.stdev(raw_estimates) if len(raw_estimates) > 1 else 1.0

        if len(raw_estimates) >= 2 and confidence > self.config.max_estimate_std:
            log.info(
                f"SKIP (low confidence): {market.question[:50]}... "
                f"std={confidence:.3f} > max={self.config.max_estimate_std:.3f}"
            )
            return None

        log.info(
            f"Estimate: {market.question[:50]}... -> {fair_prob:.2%} "
            f"(n={len(raw_estimates)}, std={confidence:.3f})"
        )

        return Estimate(
            market_condition_id=market.condition_id,
            question=market.question,
            fair_probability=fair_prob,
            raw_estimates=raw_estimates,
            confidence=confidence,
            reasoning_summary=first_reasoning,
            input_tokens_used=total_input,
            output_tokens_used=total_output,
        )

    # ── Provider dispatch ──────────────────────────────────────────────────

    def _single_call(self, market: MarketInfo):
        """Single call to the configured AI provider. Returns (prob, reasoning, in_tok, out_tok) or None."""
        if self._provider == "anthropic":
            return self._call_anthropic(market)
        elif self._provider == "gemini":
            return self._call_gemini(market)
        elif self._provider in ("openai", "openrouter", "azure_openai"):
            return self._call_openai_compat(market)
        else:
            log.error(f"Unknown AI provider: {self._provider}")
            return None

    def _parse_json_response(self, text: str):
        """Parse probability JSON from model response. Returns (prob, reasoning) or None."""
        try:
            if text.startswith("```"):
                lines = [l for l in text.split("\n") if not l.strip().startswith("```")]
                text = "\n".join(lines)
            data = json.loads(text)
            prob = float(data["probability"])
            reasoning = data.get("reasoning", "")
            prob = max(0.02, min(0.98, prob))
            return prob, reasoning
        except (json.JSONDecodeError, KeyError, ValueError) as e:
            log.debug(f"Failed to parse estimate response: {e}")
            return None

    # ── Anthropic ─────────────────────────────────────────────────────────

    def _call_anthropic(self, market: MarketInfo):
        try:
            response = self._anthropic_client.messages.create(
                model=self._model,
                max_tokens=self.config.max_estimate_tokens,
                temperature=self.config.ensemble_temperature,
                system=SYSTEM_PROMPT,
                messages=[{"role": "user", "content": _build_user_prompt(market)}],
            )
            text = response.content[0].text.strip()
            in_tok = response.usage.input_tokens
            out_tok = response.usage.output_tokens
            result = self._parse_json_response(text)
            if result is None:
                return None
            prob, reasoning = result
            return prob, reasoning, in_tok, out_tok
        except self._anthropic_sdk.RateLimitError:
            log.warning("Anthropic rate limit — waiting 5s")
            time.sleep(5)
            return None
        except self._anthropic_sdk.APIError as e:
            log.error(f"Anthropic API error: {e}")
            return None

    # ── OpenAI-compatible (OpenAI, OpenRouter, Azure OpenAI) ──────────────

    def _call_openai_compat(self, market: MarketInfo):
        provider = self._provider
        model = self._model

        if provider == "openai":
            host = (self.config.openai_api_host or "https://api.openai.com").rstrip("/")
            url = f"{host}/v1/chat/completions"
            headers = {"Authorization": f"Bearer {self.config.openai_api_key}"}
        elif provider == "openrouter":
            host = (self.config.openrouter_api_host or "https://openrouter.ai").rstrip("/")
            url = f"{host}/api/v1/chat/completions"
            headers = {"Authorization": f"Bearer {self.config.openrouter_api_key}"}
        else:  # azure_openai
            endpoint = self.config.azure_openai_endpoint.rstrip("/")
            deployment = self.config.azure_openai_deployment
            version = self.config.azure_openai_api_version or "2024-02-01"
            url = f"{endpoint}/openai/deployments/{deployment}/chat/completions?api-version={version}"
            headers = {"api-key": self.config.azure_openai_api_key}
            model = deployment  # Azure uses deployment name

        headers["Content-Type"] = "application/json"
        payload = {
            "model": model,
            "messages": [
                {"role": "system", "content": SYSTEM_PROMPT},
                {"role": "user", "content": _build_user_prompt(market)},
            ],
            "temperature": self.config.ensemble_temperature,
            "max_tokens": self.config.max_estimate_tokens,
        }

        try:
            resp = requests.post(url, json=payload, headers=headers, timeout=30)
            if resp.status_code == 429:
                log.warning(f"{provider} rate limit — waiting 5s")
                time.sleep(5)
                return None
            resp.raise_for_status()
            data = resp.json()
            text = data["choices"][0]["message"]["content"].strip()
            usage = data.get("usage", {})
            in_tok = usage.get("prompt_tokens", 0)
            out_tok = usage.get("completion_tokens", 0)
            result = self._parse_json_response(text)
            if result is None:
                return None
            prob, reasoning = result
            return prob, reasoning, in_tok, out_tok
        except requests.exceptions.HTTPError as e:
            log.error(f"{provider} API error: {e}")
            return None
        except Exception as e:
            log.debug(f"{provider} call failed: {e}")
            return None

    # ── Google Gemini ─────────────────────────────────────────────────────

    def _call_gemini(self, market: MarketInfo):
        model = self._model
        host = (self.config.gemini_api_host or "https://generativelanguage.googleapis.com").rstrip("/")
        url = f"{host}/v1beta/models/{model}:generateContent?key={self.config.gemini_api_key}"
        payload = {
            "systemInstruction": {"parts": [{"text": SYSTEM_PROMPT}]},
            "contents": [{"role": "user", "parts": [{"text": _build_user_prompt(market)}]}],
            "generationConfig": {
                "temperature": self.config.ensemble_temperature,
                "maxOutputTokens": self.config.max_estimate_tokens,
            },
        }
        try:
            resp = requests.post(url, json=payload, headers={"Content-Type": "application/json"}, timeout=30)
            if resp.status_code == 429:
                log.warning("Gemini rate limit — waiting 5s")
                time.sleep(5)
                return None
            resp.raise_for_status()
            data = resp.json()
            text = data["candidates"][0]["content"]["parts"][0]["text"].strip()
            usage = data.get("usageMetadata", {})
            in_tok = usage.get("promptTokenCount", 0)
            out_tok = usage.get("candidatesTokenCount", 0)
            result = self._parse_json_response(text)
            if result is None:
                return None
            prob, reasoning = result
            return prob, reasoning, in_tok, out_tok
        except requests.exceptions.HTTPError as e:
            log.error(f"Gemini API error: {e}")
            return None
        except Exception as e:
            log.debug(f"Gemini call failed: {e}")
            return None

    # ── API key validation ────────────────────────────────────────────────

    def validate_api_key(self) -> bool:
        """Make a minimal test call to validate the configured provider's API key.

        Returns False on auth failure (HTTP 401/403, AuthenticationError).
        Other errors (network, rate-limit) return True so transient failures don't block startup.
        """
        provider = self._provider
        try:
            if provider == "anthropic":
                self._anthropic_client.messages.create(
                    model=self._model,
                    max_tokens=1,
                    messages=[{"role": "user", "content": "hi"}],
                )
                return True

            elif provider in ("openai", "openrouter", "azure_openai"):
                if provider == "openai":
                    host = (self.config.openai_api_host or "https://api.openai.com").rstrip("/")
                    url = f"{host}/v1/chat/completions"
                    auth_headers = {"Authorization": f"Bearer {self.config.openai_api_key}"}
                elif provider == "openrouter":
                    host = (self.config.openrouter_api_host or "https://openrouter.ai").rstrip("/")
                    url = f"{host}/api/v1/chat/completions"
                    auth_headers = {"Authorization": f"Bearer {self.config.openrouter_api_key}"}
                else:
                    endpoint = self.config.azure_openai_endpoint.rstrip("/")
                    deployment = self.config.azure_openai_deployment
                    version = self.config.azure_openai_api_version or "2024-02-01"
                    url = f"{endpoint}/openai/deployments/{deployment}/chat/completions?api-version={version}"
                    auth_headers = {"api-key": self.config.azure_openai_api_key}
                resp = requests.post(
                    url,
                    json={"model": self._model, "messages": [{"role": "user", "content": "hi"}], "max_tokens": 1},
                    headers={**auth_headers, "Content-Type": "application/json"},
                    timeout=10,
                )
                if resp.status_code in (401, 403):
                    return False
                return True

            elif provider == "gemini":
                host = (self.config.gemini_api_host or "https://generativelanguage.googleapis.com").rstrip("/")
                resp = requests.post(
                    f"{host}/v1beta/models/{self._model}:generateContent"
                    f"?key={self.config.gemini_api_key}",
                    json={"contents": [{"parts": [{"text": "hi"}]}], "generationConfig": {"maxOutputTokens": 1}},
                    headers={"Content-Type": "application/json"},
                    timeout=10,
                )
                # Gemini returns 400 with "API key" in body for invalid keys (not 401)
                if resp.status_code == 403:
                    return False
                if resp.status_code == 400 and "API key" in resp.text:
                    return False
                return True

        except Exception as e:
            if self._anthropic_sdk and isinstance(e, self._anthropic_sdk.AuthenticationError):
                return False
            # Network errors, rate limits, etc. — don't block startup
            return True

        return True
